### DISCUSSION TITLE
Pattern for where to parse function input, if you need input in multiple places in zod?

### DISCUSSION THREAD

**Discussion by lancejpollard** [colinhacks/zod#3159] [Q&A] [answered]
I have a codebase which has basically this organization:

```ts
export function handleOnCLI(input) {
  // need it at least partially parsed here.
  if (input.blah === 10) {
    convertFile(input)
  }
  // ... tons more
}

export function convertFile(input) {
  // need it parsed here fully
  const cmd = buildCommand(input)
  exec(cmd)
}

export function buildCommand(input) {
  // need it parsed here fully
  return `my-command ${input.arg1} ${input.arg2}`
}
```

The thing is, all of these methods are publicly callable:

1. You can call the CLI with serialized args.
2. You can cal `convertFile` with your own input.
3. You can call `buildCommand` to render blog posts. Etc..

So it seems that I should place the `zod` parser/schema/model in all 3 places, like this:

```ts
export function handleOnCLI(source) {
  const input = MySchema.parse(source)
  // need it at least partially parsed here.
  if (input.blah === 10) {
    convertFile(input)
  }
  // ... tons more
}

export function convertFile(source) {
  const input = MySchema.parse(source)
  // need it parsed here fully
  const cmd = buildCommand(input)
  exec(cmd)
}

export function buildCommand(source) {
  const input = MySchema.parse(source)
  // need it parsed here fully
  return `my-command ${input.arg1} ${input.arg2}`
}
```

But, obviously this means my (potentially large) data objects are being parsed 3 times on every call which is a huge waste of performance.

So at first I started putting them in the lowest-level function (`buildCommand`), but that resulted in me doing all kinds of inference typechecks on the outer functions, which was ugly. So then I moved it out one to the `convertFile` function. But sometimes I need to initialize the `input` even before that. So I'm thinking of moving it out all the way "to the top". But there is no top!

So it's like, my API needs to require people to do 2 things for every call. So instead of 1-liners:

```ts
convertFile(source)
```

I need 2-liners everywhere:

```ts
const input = MySchema.parse(source)
convertFile(input)
```

But this also means you can _forget_ to call `MySchema.parse` and pass in unsanitized input accidentally. Ideally not, since you will be developing/debugging for a while before it's released, but still. Having to call parse right before each method call feels cumbersome. But it beats parsing the input multiple times.

Another option I thought of was to pass a flag as the 2nd arg to each function, like `thisHasBeenParsed`. But still problems with that.

Calling it manually every time gives you extra power though because you know what you're getting, and every nested function can have the TypeScript type for that schema, so at least in dev you won't make the mistake.

What are your thoughts on this? Is there any better approach?

**Comment by JacobWeisenburger** [comment, marked as answer]
Would you be able to give us some example shell inputs of how these functions are being called?

Something like:
```
$ myCustomCli arg1 arg2
```

Could you also give the definition of `MySchema`?
